src/eduid_common/config/base.py

```diff
@@ -0,0 +1,133 @@
+    broker_transport_options: dict = field(default_factory=lambda: {"fanout_prefix": True})
+    task_routes: dict = field(default_factory=lambda: {
+      'eduid_am.*': {'queue': 'am'},
+      'eduid_msg.*': {'queue': 'msg'},
+      'eduid_letter_proofing.*': {'queue': 'letter_proofing'}})
+
+
+@dataclass(frozen=True)
+class BaseConfig(object):
+    devel_mode: bool = False
+    development: bool = False
+    # enable disable debug mode
+    debug: bool = False
+    # session cookie
+    session_cookie_name: str = 'sessid'
+    session_cookie_path: str = '/'
+    session_cookie_timeout: int = 60      # in minutes
+    # the domain for the session cookie. If this is not set, the cookie will
+    # be valid for all subdomains of SERVER_NAME.
+    session_cookie_domain: str = ''
+    session_cookie_secure: bool = True
+    session_cookie_httponly: bool = False
+    # The URL scheme that should be used for URL generation if no URL scheme is
+    # available. This defaults to http
+    preferred_url_scheme: str = 'https'
+    # Redis config
+    # The Redis host to use for session storage.
+    redis_host: Optional[str] = None
+    # The port of the Redis server (integer).
+    redis_port: int = 6379
+    # The Redis database number (integer).
+    redis_db: int = 0
+    # Redis sentinel hosts, comma separated
+    redis_sentinel_hosts: Optional[List[str]] = None
+    # The Redis sentinel 'service name'.
+    redis_sentinel_service_name: str = 'redis-cluster'
+    saml2_logout_redirect_url: str = 'https://eduid.se/'
+    token_service_url: str = ''
+    celery_config: CeleryConfig = field(default_factory=CeleryConfig)
+    available_languages: Dict[str, str] = field(default_factory=lambda: {
+        'en': 'English',
+        'sv': 'Svenska'
+        })
+    mail_default_from: str = 'info@eduid.se'
+    eduid_site_url: str = ''
+    eduid_static_url: str = ''
+    static_url: str = ''
+    dashboard_url: str = ''
+    reset_passwd_url: str = ''
+    # set absolute URL so it can be included in emails
+    signup_url: str = ''
+    no_authn_urls: list = field(default_factory=lambda: [
+        "^/status/healthy$",
+        "^/status/sanity-check$"
+        ])
+    action_plugins: list = field(default_factory=lambda: [
+        "tou",
+        "mfa"
+        ])
+    # environment=(dev|staging|pro)
+    environment: str = 'dev'
+    tou_version: str = '2017-v6'
+    current_tou_version: str = '2017-v6'  # backwards compat
+    fido2_rp_id: str = ''
+    secret_key: str = ''
+
+    def __post_init__(self):
+        if isinstance(self.celery_config, dict):
+            object.__setattr__(self, 'celery_config', CeleryConfig(**self.celery_config))
+
+    def __getitem__(self, attr: str) -> Any:
+        '''
+        XXX Once the apps are all accessing configuration as attributes,
+            we will be able to remove this method.
+        '''
+        try:
+            return self.__getattribute__(attr)
```

Comment:
Don't we need attr.lower() to be backwards compatible with the other apps?

Reply:
The objective is to finish up with all webapps accessing config keys as lower case (typed) attributes of the config object. So  my idea is that when the webapps start using these config classes, it will happen within a migration not just to start using them, but also use them as used in the IdP - as lower case attributes.
Depending on how we plan that migration, we may or may not need an intermediate step where access is lower cased.
But keeping permanently settings access through __gettitem__ or __getattribute__ would prevent mypy from detecting type errors.

Reply:
Do we need the __getitem__ at all then if not to make this config object compatible with the current way our apps look up config? I mean if we need to change the uppercase keys to lowercase we can just as well switch to attributes directly.

Reply:
I think we agree, what I was proposing above was to migrate the webapps directly to lower case attributes.
We atm need __getitem__ both for the webapps and also for cherrypy, which accesses the settings as lower case dict keys, so as soon as we migrate away from cherrypy and use attr access in the webapps, we can get rid of it.